Make the docker proxy a standalone binary not a re-exec

justincormack opened the pull request:
This reduces memory usage with a lot of docker proxy processes.
On Docker for Mac we are currently carrying a patch to replace
the binary as we modify it to forward ports to the Mac rather
than the Linux VM, this allows us to simply replace this binary
in our packaging with one that has a compatible interface. This
patch does not provide an easy way to substitute a binary as
the interface is complex and there are few use cases, but where
needed this can be done.

Signed-off-by: Justin Cormack justin.cormack@docker.com

Replaces #21746 cc @djs55

![littlerobin](https://cloud.githubusercontent.com/assets/482364/15828371/f651e404-2c06-11e6-9112-88883af5a42c.jpg)

justincormack committed:
Make the docker proxy a standalone binary not a re-exec

This reduces memory usage with a lot of docker proxy processes.
On Docker for Mac we are currently carrying a patch to replace
the binary as we modify it to forward ports to the Mac rather
than the Linux VM, this allows us to simply replace this binary
in our packaging with one that has a compatible interface. This
patch does not provide an easy way to substitute a binary as
the interface is complex and there are few use cases, but where
needed this can be done.

Signed-off-by: Justin Cormack <justin.cormack@docker.com>

icecrime commented:
Legitimate failure on win2lin?

```
11:06:05 # github.com/docker/docker/cmd/proxy
11:06:05 cmd\proxy\main.go:60: undefined: syscall.SIGSTOP
```

icecrime commented:
Otherwise totally +1 on design :+1: I think we can reasonably go to code review as it's something we talked about in the past with @crosbymichael and @tonistiigi.

crosbymichael commented on cmd/proxy/main.go:
```diff
@@ -0,0 +1,67 @@
+)
+
+func main() {
+	f := os.NewFile(3, "signal-parent")
```
I don't understand what this is for?

crosbymichael commented on cmd/proxy/main.go:
```diff
@@ -0,0 +1,67 @@
+
+func handleStopSignals(p Proxy) {
+	s := make(chan os.Signal, 10)
+	signal.Notify(s, os.Interrupt, syscall.SIGTERM, syscall.SIGSTOP)
```
you can remove SIGSTOP here as you cannot handle it anyways to fix the failures in win2lin

thaJeztah commented:
This is weird, looks like a test is failing, but didn't on the PR that updated it;

```
11:41:09 ----------------------------------------------------------------------
11:41:09 FAIL: docker_cli_rename_test.go:69: DockerSuite.TestRenameInvalidName
11:41:09 
11:41:09 docker_cli_rename_test.go:78:
11:41:09     c.Assert(out, checker.Contains, "\"docker rename\" requires exactly 2 argument(s).", check.Commentf("%v", err))
11:41:09 ... obtained string = "Error: Neither old nor new names may be empty\n"
11:41:09 ... substring string = "\"docker rename\" requires exactly 2 argument(s)."
11:41:09 ... exit status 1: Error: Neither old nor new names may be empty
11:41:09 
11:41:09 
11:41:10 
```

thaJeztah commented:
Genuine failures on win2lin it looks like;

```
11:25:14 ----------------------------------------------------------------------
11:25:14 FAIL: docker_cli_build_test.go:2897: DockerSuite.TestBuildAddLocalAndRemoteFilesWithoutCache
11:25:14 
11:25:14 docker_cli_build_test.go:2905:
11:25:14     c.Fatal(err)
11:25:14 ... Error: failed to start file storage container. ec=127
11:25:14 out=9efbf114f222bd20b5d32766b0ffb61bbb7d6e4f4985ce4cb446dc634188425c
11:25:14 C:\CI\CI-cf0c516\docker.exe: Error response from daemon: driver failed programming external connectivity on endpoint fileserver-cnt-bvijpnaizt (da74bdbd7cc532e0bbcf299b42ead9e75922584ef68e7eb8d0059baba1d81af7): exec: "docker-proxy": executable file not found in $PATH.
11:25:14 
11:25:14 err=exit status 127
```

justincormack commented:
Still trying to work out why win2lin doesn't find the binaries, will try a different temporary debug commit...

justincormack commented:
Ok I think the win2lin script will need to change to copy with `docker-proxy` into the PATH, not just `dockerd`.

justincormack commented:
vendor expected to fail right now, will make a vendor PR. win2lin needs to be modified to copy the docker-proxy into the build at https://github.com/Microsoft/docker/blob/9ea3f2aca2ce3b12d42e72290e404592209ee652/hack/Jenkins/W2L/setup.sh#L199

justincormack commented:
libnetwork vendoring PR is https://github.com/docker/libnetwork/pull/1230

runcom commented:
Are there any benchmarks showing the memory gain with this?  This is awesome though!

thaJeztah commented:
ping @justincormack looks like this needs a rebase now :cry:

justincormack commented:
Reworking it to make it able to pass the libnetwork CI... Will update in a
bit...
On 13 Jun 2016 6:21 p.m., "Sebastiaan van Stijn" <notifications@github.com>
wrote:

> ping @justincormack <https://github.com/justincormack> looks like this
> needs a rebase now 😢
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/docker/docker/pull/23312#issuecomment-225649059>, or mute
> the thread
> <https://github.com/notifications/unsubscribe/AAdcPHx9fyu7v4aoVU_FdePZrKpztRLyks5qLZGtgaJpZM4IvCs7>
> .
>

justincormack commented:
Ok, rebased and changed the vendoring so libnetwork tests should be passing now - most of the code moved into libnetwork, which I think makes more sense. Hopefully this all works now...

justincormack commented:
win2lin needs the following patch applied to setup.sh in CI to copy the `docker-proxy` binary:
(current script is https://github.com/docker/docker/blob/master/hack/Jenkins/W2L/setup.sh)

```
--- setup.sh.orig	2016-06-14 00:52:49.000000000 +0100
+++ setup.sh	2016-06-14 00:54:58.000000000 +0100
@@ -197,6 +197,7 @@
 FROM docker:$COMMITHASH
 RUN hack/make.sh binary
 RUN cp bundles/latest/binary-daemon/dockerd /bin/dockerd 
+RUN cp bundles/latest/binary-daemon/docker-proxy /bin/docker-proxy
 CMD dockerd -D -H tcp://0.0.0.0:$port_inner $daemon_extra_args
 EOF
 
```

icecrime commented:
We should merge it soon because it's a really cool feature, but I don't think it'll make the 1.12.0 code freeze deadline :disappointed: Removing from the milestone, sorry.

justincormack commented:
As pointed out on https://github.com/docker/libnetwork/pull/1230#issuecomment-225917125 the vendoring code will not pull in code to build binaries from libnetwork. Anyone got preferences on how to vendor in this situation?

tiborvass commented:
I'm tentatively adding this back to the 1.12 milestone since it's needed by docker for desktop.

justincormack commented:
Ok I rebased and modified the vendoring code to keep the proxy command imported form libnetwork, so this should vendor without issues now.

cc @tiborvass

tiborvass commented:
@justincormack still some issues on the vendoring. Let us know if you need help

justincormack commented:
@tiborvass well I need https://github.com/docker/libnetwork/pull/1230 merged to update the libnetwork vendoring

thaJeztah commented:
@justincormack https://github.com/docker/libnetwork/pull/1230 was just merged

mavenugo commented:
@justincormack @tiborvass @thaJeztah https://github.com/docker/libnetwork/pull/1230 is merged. But we also need https://github.com/docker/swarmkit/pull/1130 to be merged and vendored in along with libnetwork vendor-in. Otherwise swarm-mode will have issues.

mavenugo commented:
@justincormack you can vendor-in libnetwork along with swarmkit (16fa595d3b6fec012830179dc8e9b2d90335527d) to pull in the user-land proxy fix that you pushed.
(along with the fixes required for https://github.com/docker/docker/issues/23990)

justincormack committed:
Update libnetwork vendoring

Update now that the changes in https://github.com/docker/libnetwork/pull/1230
have been merged

Signed-off-by: Justin Cormack <justin.cormack@docker.com>

justincormack commented:
Ok, pushing these changes, will squash shortly.

justincormack committed:
Update swarmkit vendoring

Needed for libnetwork vendoring

Update Secret API name change correspondingly

Signed-off-by: Justin Cormack <justin.cormack@docker.com>

mavenugo commented:
@justincormack thanks. i dont see a need to squash these. having independent vendoring makes it easy for review.

The CI failures seems genuine and needs corresponding fixes in engine to handle swarmkit changes.

justincormack commented:
I think we need to vendor the swarmkit changes and the libnetwork changes without the final proxy commit in with the corresponding changes to docker where it uses swarmapi first, and then just update ti the final libnetwork proxy commit here. cc @tiborvass

justincormack commented:
@mavenugo I think we were both commenting at the same time. Yes, it is swarmkit api changes in engine causing the failures...

mavenugo commented:
@justincormack due to the way vendor-in works (and the order of the commits matter on the dependent projects), we have to think about a different strategy. 
* Either we can resolve build issue in this commit and take in all the vendoring together (preferred)
* Identify the appropriate commit in swarmkit that introduced the API change and get that vendored in first. Then followed by pulling in latest swarmkit SHA & libnetwork.

justincormack commented:
Hmm, so the swarmkit update seems to be causing a real failure with 
```
16:35:34 
16:35:34 ----------------------------------------------------------------------
16:35:34 FAIL: docker_api_swarm_test.go:248: DockerSwarmSuite.TestApiSwarmPromoteDemote
16:35:34 
16:35:34 [d4303199] waiting for daemon to start
16:35:34 [d4303199] daemon started
16:35:34 [d16803067] waiting for daemon to start
16:35:34 [d16803067] daemon started
16:35:34 docker_api_swarm_test.go:288:
16:35:34     d1.updateNode(c, d2.NodeID, func(n *swarm.Node) {
16:35:34         n.Spec.Role = swarm.NodeRoleWorker
16:35:34     })
16:35:34 github.com/docker/docker/integration-cli/_test/_obj_test/daemon_swarm.go:249:
16:35:34     ...open github.com/docker/docker/integration-cli/_test/_obj_test/daemon_swarm.go: no such file or directory
16:35:34 ... obtained int = 500
16:35:34 ... expected int = 200
16:35:34 ... output: "{\"message\":\"rpc error: code = 5 desc = can't find manager in raft memberlist\"}\n"
16:35:34 
16:35:34 [d4303199] exiting daemon
16:35:34 [d16803067] exiting daemon
16:35:58 
```

Anyone know what is up with this?

cpuguy83 commented:
ping @aaronlehmann @aluzzardi

tonistiigi commented:
@abronan ^

tonistiigi commented:
@abronan @aaronlehmann If I add a 1 sec sleep before the demotion in https://github.com/docker/docker/blob/6167a9ab166ce1e6fd0ffa653b9476d33983f15b/integration-cli/docker_api_swarm_test.go#L286 the test starts to pass. So even though the manager appears active, for a short amount of time it can't be demoted.

aaronlehmann commented:
Makes sense given the code in https://github.com/docker/swarmkit/pull/1091. It won't let you demote until the manager is part of the raft cluster. @abronan I think this case should not return an error. If the manager is not part of raft, it should still be possible to demote it. However I'm not sure how we would deal with the race condition of a promotion going through at the same time as we demote.

tonistiigi commented:
@aaronlehmann The test waits for the control API to become active, normally this should not happen before the manager is part of cluster and leader has been determined.

tonistiigi commented:
Seems no problem with demotion, the test has a bug. @justincormack There should be no else block in https://github.com/docker/docker/blob/6167a9ab166ce1e6fd0ffa653b9476d33983f15b/integration-cli/docker_api_swarm_test.go#L280 https://github.com/docker/docker/blob/6167a9ab166ce1e6fd0ffa653b9476d33983f15b/integration-cli/docker_api_swarm_test.go#L299

justincormack commented:
@tonistiigi thanks!

justincormack committed:
Fix test case for docker_api_swarm_test.go

Signed-off-by: Justin Cormack <justin.cormack@docker.com>

tonistiigi commented:
LGTM

justincormack commented:
All green as tp5 decommissioned.

tiborvass commented:
LGTM

thaJeztah commented:
We'll do a docs follow-up PR for this

tswift242 commented:
@thaJeztah Looks like this wasn't included in the 1.12.0 changelog? Also looks like the docs haven't been updated yet.

thaJeztah commented:
@tswift242 thanks for the ping, you're right! Would you mind opening an issue?

tswift242 commented:
@thaJeztah Opened https://github.com/docker/docker/issues/25686
